File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    public int numIslands(char[][] grid) 
    {
        int n = grid.length;
        int m = grid[0].length;
        boolean visited[][] = new boolean[n][m];
        int count = 0;
        for(int i = 0;i<grid.length;i++){
            for(int j = 0;j<grid[i].length;j++){
                if(!visited[i][j] && grid[i][j] == '1'){
                    dfs(grid,visited,i,j);
                    count++;
                }
            }
        }
        return count;
    }
    public void dfs(char[][] grid, boolean visited[][], int i, int j){
        visited[i][j] = true;
        int directions[][] = new int[][]{{-1,0},{1,0},{0,-1},{0,1}};
        for(int[] direction : directions){
            int x = i + direction[0];
            int y = j + direction[1];
            if(x<0 || y<0 || x>=grid.length || y>=grid[i].length || visited[x][y] || grid[x][y] == '0'){
                continue;
            }
            else{
                dfs(grid,visited,x,y);
            }
        }
    }
}

File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.py
class Solution:
    def helper(self, grid, i, j, n, m, visited):
        visited[i][j] = True
        directions = [(1,0),(-1,0),(0,1),(0,-1)]
        for x, y in directions:
            new_x = i+x
            new_y = j+y
            if new_x >=0 and new_y >= 0 and new_x < n and new_y < m and not visited[new_x][new_y] and grid[new_x][new_y] >= grid[i][j]:
                self.helper(grid, i+x, j+y, n, m, visited)

    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        n = len(heights)
        m = len(heights[0])
        pacific_visited = [[False]*m for _ in range(n)]
        atlantic_visited = [[False]*m for _ in range(n)]
        #top_row
        for i in range(m):
            # pacific_visited[0][i] = True
            self.helper(heights,0,i,n,m,pacific_visited)
        
        #left_col
        for i in range(n):
            # pacific_visited[i][0] = True
            self.helper(heights,i,0,n,m,pacific_visited)
        
        #bottom_row
        for i in range(m):
            self.helper(heights,n-1,i,n,m,atlantic_visited)

        #right_col
        for i in range(n):
            self.helper(heights,i,m-1,n,m,atlantic_visited)
        
        ans = []
        for i in range(n):
            for j in range(m):
                if pacific_visited[i][j] and atlantic_visited[i][j]:
                    ans.append([i,j])
        return ans



        